### PR TITLE
Handle Avro schema generation for logical data types in BigQueryAvroUtils

### PR DESCRIPTION
**Pull request by kkdoon**
As discussed in issue #23541, `BigQueryAvroUtils` does not correctly generate the Avro schema for certain BigQuery data types, such as `Numeric` and `BigNumeric`. This PR address this issue and generates the correct avro schema. This further helps in transformation of GenericRecord -> TableRow, via `convertGenericRecordToTableRow` function, that rely on these BigQuery data types.

R: @steveniemitz @kennknowles @vachan-shetty

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kkdoon** [comment]
assign set of reviewers

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]
Just one change needed but otherwise LGTM, thanks for following through with this!

**Comment by apilloud** [comment]
R: @steveniemitz @kennknowles @vachan-shetty

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOReadTest.java
Comment (kkdoon) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOReadTest.java]: sounds good, i have updated the this test.

**Comment by damccorm** [comment]
@ahmedabu98 what are next steps here? Looks like we may just need a committer review - @apilloud could you take a look?

**Review by ahmedabu98 — APPROVED** [review]
@damccorm yup, this LGTM

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java
Comment (apilloud) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:466]: The documentation you linked says `76.76` and the API docs say `1 <= precision - scale <= 38 and 0 <= scale <= 38` which means the max value for precision is `76`. I'm not sure what will work in practice it seems like `76` would potentially fail to read from BigQuery and `77` will fail to write to BigQuery. Can you provide any more info on why 77 was chosen? Should there be tests for this edge case?
https://cloud.google.com/java/docs/reference/google-cloud-bigquerystorage/2.8.2/com.google.cloud.bigquery.storage.v1.TableFieldSchema.Builder#com_google_cloud_bigquery_storage_v1_TableFieldSchema_Builder_getPrecision__

**Review by apilloud — APPROVED** [review]
LGTM

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java
Comment (kkdoon) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:466]: I ran a beam job and tried reading from an actual BigQuery table with BIGNUMERIC field. The [writer schema](https://github.com/apache/beam/blob/483a0c95734c528aa45419596e9f27e9e650c5d7/sdks/java/core/src/main/java/org/apache/beam/sdk/io/AvroSource.java#L754), set in the extracted Avro file, has precision value set as 77 and the scale is set to 38, for fields that have no explicit precision/scale specified.  

I also tried inserting the max supported value that has a precision of 77 and it works as long as its <= 578960446186580977117854925043439539266.34992332820282019728792003956564819967 (since 77th digit is partial it supports this value and not 999999999999999999999999999999999999999.99999999999999999999999999999999999999 instead)

Comment (apilloud) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:466]: The tests you describe don't seem to be in the code and this didn't answer my question but I guess the answer is in the initial question. A value of `77` will at least pull the failure into something the user can fix with additional constraints within their pipeline.

Comment (kkdoon) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryAvroUtils.java:466]: sorry, i was not referring to these unit tests. I tested with a standalone beam job, in direct runner mode, against an actual BigQuery table and verified the default precision/scale value set in the Avro writer schema (that's why chose to set default as 77 instead of 76). Also, tried inserting and reading the max documented value for BigNumeric field and it works for numbers with 77 precision. I could add a BQ integration test to test this edge case, if that helps.

**Pull request merged**
